contrib/pyln-proto/pyln/proto/wire.py

```diff
@@ -0,0 +1,395 @@
+
+    def handshake_act_three_initiator(self):
+        h = Sha256Mixer(b'')
+        h.hash = self.handshake['h']
+        pk = self.local_pubkey.serializeCompressed()
+        c = encryptWithAD(self.temp_k2, self.nonce(1), h.digest(), pk)
+        h.update(c)
+        se = ecdh(self.local_privkey, self.re)
+
+        self.chaining_key, self.temp_k3 = hkdf_two_keys(
+            salt=self.chaining_key, ikm=se.raw
+        )
+        t = encryptWithAD(self.temp_k3, self.nonce(0), h.digest(), b'')
+        m = b'\x00' + c + t
+        t = hkdf(salt=self.chaining_key, ikm=b'', info=b'')
+
+        self.sk, self.rk = hkdf_two_keys(salt=self.chaining_key, ikm=b'')
+        self.rn, self.sn = 0, 0
+        return m
+
+    def handshake_act_three_responder(self, m):
+        h = Sha256Mixer(b'')
+        h.hash = self.handshake['h']
+        v, c, t = m[0], m[1:50], m[50:]
+        if v != 0:
+            raise ValueError("Unsupported handshake version {}, only version "
+                             "0 is supported.".format(v))
+        rs = decryptWithAD(self.temp_k2, self.nonce(1), h.digest(), c)
+        h.update(c)
+        se = ecdh(self.handshake['e'], PublicKey(rs))
+
+        self.chaining_key, self.temp_k3 = hkdf_two_keys(
+            se.raw, self.chaining_key
+        )
+        decryptWithAD(self.temp_k3, self.nonce(0), h.digest(), t)
+        self.rn, self.sn = 0, 0
+
+        self.rk, self.sk = hkdf_two_keys(salt=self.chaining_key, ikm=b'')
+
+    def read_message(self):
+        with self.recv_lock:
+            lc = self.connection.recv(18)
+            if len(lc) != 18:
+                raise ValueError(
+                    "Short read reading the message length: 18 != {}".format(
+                        len(lc))
+                )
+            length = decryptWithAD(self.rk, self.nonce(self.rn), b'', lc)
+            length, = struct.unpack("!H", length)
+            self.rn += 1
+
+            mc = self.connection.recv(length + 16)
+            if len(mc) < length + 16:
+                raise ValueError(
+                    "Short read reading the message: {} != {}".format(
+                        length + 16, len(lc)
+                    )
+                )
+            m = decryptWithAD(self.rk, self.nonce(self.rn), b'', mc)
+            self.rn += 1
+            assert(self.rn % 2 == 0)
+            self._maybe_rotate_keys()
+
+        return m
+
+    def send_message(self, m):
+        length = struct.pack("!H", len(m))
+        with self.send_lock:
+            lc = encryptWithAD(self.sk, self.nonce(self.sn), b'', length)
+            mc = encryptWithAD(self.sk, self.nonce(self.sn + 1), b'', m)
+            self.sn += 2
+            self.connection.send(lc)
+            self.connection.send(mc)
+            assert(self.sn % 2 == 0)
+            self._maybe_rotate_keys()
+
+    def _maybe_rotate_keys(self):
+        if self.sn == 1000:
+            self.sck, self.sk = hkdf_two_keys(salt=self.sck, ikm=self.sk)
```

Comment:
only the tests `rotation` and `shake` set `self.sck`. if i expand `test_handshake` to send more, i get an exception as `self.sck` is not set.

Reply:
I didn't intend for `test_handshake` to leave `lc1` and `lc2` in a usable state, since they just step through the test vector to pin-point an eventual regression. I added the initialization of `sck` and `rck` anyway :-)